### PR TITLE
perf(sampling): Optimize top_k_top_p_sampling_from_logits/from_probs for large-vocab small-k sampling

### LINKED ISSUE #3389: [Sampling Perf] top_k_top_p_sampling on SM 120 is significantly slower than naive insertion sort-based speclialized kernel
I noticed that FlashInfer's `top_k_top_p_sampling_from_logits` shows lower performance compared to a naive register-level insertion sort kernel when benchmarked on the SM 120 architecture (specifically, a laptop RTX 5060).

Even with real-world distribution data (tested on Qwen3-8B sampled logits), a straightforward custom kernel using a template-specialized, unroll-forced insertion sort out-performs FlashInfer across various batch sizes.

I want to raise this discussion to see if this is an expected architectural trade-off due to FlashInfer's dynamic generalization (rejection sampling approach) or if there are specific tuning parameters for consumer-grade Blackwell GPUs (SM 120) that I missed.

```yaml
####################################################################################################
bs: 1, vocab_size: 256000
torch                                    mean time: 0.519055 ms, 0.99 GB/s
flashinfer                               mean time: 0.140417 ms, speedup: 3.70, 3.65 GB/s
sampling_topk_topp_batched               mean time: 0.084192 ms, speedup: 6.17, 6.08 GB/s
sampling_topk_topp_split_k               mean time: 0.036327 ms, speedup: 14.29, 14.09 GB/s
####################################################################################################
bs: 4, vocab_size: 256000
torch                                    mean time: 0.931264 ms, 2.20 GB/s
flashinfer                               mean time: 0.154697 ms, speedup: 6.02, 13.24 GB/s
sampling_topk_topp_batched               mean time: 0.084510 ms, speedup: 11.02, 24.23 GB/s
sampling_topk_topp_split_k               mean time: 0.051476 ms, speedup: 18.09, 39.79 GB/s
####################################################################################################
bs: 8, vocab_size: 256000
torch                                    mean time: 1.296351 ms, 3.16 GB/s
flashinfer                               mean time: 0.182281 ms, speedup: 7.11, 22.47 GB/s
sampling_topk_topp_batched               mean time: 0.084636 ms, speedup: 15.32, 48.40 GB/s
sampling_topk_topp_split_k               mean time: 0.075195 ms, speedup: 17.24, 54.47 GB/s
####################################################################################################
bs: 16, vocab_size: 256000
torch                                    mean time: 2.317547 ms, 3.53 GB/s
flashinfer                               mean time: 0.307686 ms, speedup: 7.53, 26.62 GB/s
sampling_topk_topp_batched               mean time: 0.102009 ms, speedup: 22.72, 80.31 GB/s
####################################################################################################
bs: 32, vocab_size: 256000
torch                                    mean time: 4.632301 ms, 3.54 GB/s
flashinfer                               mean time: 0.598686 ms, speedup: 7.74, 27.37 GB/s
sampling_topk_topp_batched               mean time: 0.127663 ms, speedup: 36.29, 128.34 GB/s
```
Reproduction and benchmark results can be found on [here](https://github.com/WingEdge777/vitamin-cuda/tree/main/kernels/sampling)
Maybe I did the benchmark in a wrong way, so any correction or review on the evaluation pipeline would be highly appreciated.

## Environments
GPU: NVIDIA GeForce RTX 5060 Laptop (SM 120)
CUDA: 13.2
flashinfer-python：0.6.9

## My Analysis & Guess
I guess the gap comes from a few architectural differences:
- Template Specialization vs. Dynamic Generalization : In my implementation, `top_k` is specialized as a template parameter. nvcc can resolve loop boundaries at compile time, leading to aggressive branchless unrolling and scalar replacement of aggregates (SRA) on register files. FlashInfer, as a general library, has to support dynamic/arbitrary top_k which inevitably introduces dynamic branches or state machines.
- Hardware Specific Tuning: FlashInfer's dual-pivot/rejection sampling might be highly tuned for datacenter GPUs (like H100/B200) but hasn't been optimized for SM 120 consumer architectures yet.

Is this behavior expected for FlashInfer's sampling algorithm when handling small top_k or when deployed on smaller laptop GPUs? Are there plans to introduce template-specialized sampling APIs for specific constraints?

Thanks for this amazing project! Looking forward to any insights.

### PR DESCRIPTION
**Pull request by bkryu**
### Summary
Adds a faster `filter_apply_order="top_k_first"` path for `top_k_top_p_sampling_from_logits` and `top_k_top_p_sampling_from_probs`. For modest `top_k` over a large vocabulary, instead of masking/renormalizing the full vocab and running rejection sampling across it, we select the top-k entries with the parallel radix top-k kernel and run top-p **over only those k survivors**. This is distribution-equivalent to the existing path but 2–4× faster for `from_logits` (see benchmarks).

Addresses the slow small-batch top-k/top-p sampling reported in #3389.

### Motivation

The existing `top_k_first` path keeps everything in full-vocab layout:
- `from_logits`: `top_k_mask_logits` → `torch.softmax` (full vocab) → `top_p_sampling_from_probs` (single-CTA rejection over full vocab)
- `from_probs`: `top_k_renorm_probs` → `top_p_sampling_from_probs` (single-CTA rejection over full vocab)

After top-k, only `k` entries are relevant, but the general full-vocab kernels still process all `V` elements — a full-vocab softmax (logits) and a multi-round single-CTA rejection scan (both). At small batch the rejection kernel launches one CTA per request and underutilizes the GPU.

### Changes

A gated fast path in `flashinfer/sampling.py` (no kernel changes):

1. Select top-k via `flashinfer.top_k(..., sorted=True)` (parallel radix selection, returns the `k` values + their indices).
2. Normalize over the `k` survivors — `softmax` (logits) or renorm (probs); these are mathematically identical to the masked full-vocab versions.
3. Run `top_p_sampling_from_probs` over the `k`-element distribution, then map the local choice back to the global vocab index via `gather`.
Both entry points route through the same helper with `sorted=True`, so they reduce to the same `probs_k` and stay **sample-aligned** with each other.

### Accuracy
- **Distribution-equivalent** to the original `top_k_first` path (validated: total-variation distance ≈ 0.007 vs. the analytic target over 40k draws).
- **Per-seed sample values may differ** from previous versions (the RNG now maps over a `k`-element domain, and radix top-k may break k-th-boundary ties differently). The two APIs remain aligned with each other.
- **CUDA-graph compatible**: the default vectorized top-k captures into a graph; the graph-safe fallback is only requested for the non-deterministic cluster path.

### Benchmarks
  
`top_k=50`, `top_p=0.9`, CUDA-graph mode, median ms. "before" = original path, "after" = fast path.

### B200 (SM100)

| API | dtype | bs | vocab | before | after | speedup |
|---|---|---|---|---|---|---|
| from_logits | bf16 | 1 | 128k | 0.121 | 0.059 | 2.05× |
| from_logits | bf16 | 8 | 128k | 0.139 | 0.064 | 2.17× |
| from_logits | bf16 | 32 | 128k | 0.147 | 0.066 | 2.23× |
| from_logits | bf16 | 256 | 128k | 0.431 | 0.180 | 2.39× |
| from_logits | bf16 | 1 | 256k | 0.206 | 0.067 | 3.07× |
| from_logits | bf16 | 32 | 256k | 0.294 | 0.075 | 3.92× |
| from_logits | fp32 | 1 | 128k | 0.126 | 0.057 | 2.21× |
| from_logits | fp32 | 8 | 128k | 0.131 | 0.063 | 2.08× |
| from_logits | fp32 | 32 | 128k | 0.163 | 0.068 | 2.40× |
| from_logits | fp32 | 256 | 128k | 0.453 | 0.122 | 3.71× |
| from_logits | fp32 | 1 | 256k | 0.208 | 0.061 | 3.41× |
| from_logits | fp32 | 32 | 256k | 0.296 | 0.113 | 2.62× |
| from_probs | fp32 | 1 | 128k | 0.076 | 0.064 | 1.19× |
| from_probs | fp32 | 8 | 128k | 0.077 | 0.067 | 1.15× |
| **from_probs** | fp32 | 32 | 128k | 0.088 | 0.105 | **0.84×** |
| from_probs | fp32 | 256 | 128k | 0.365 | 0.155 | 2.35× |
| from_probs | fp32 | 1 | 256k | 0.099 | 0.067 | 1.48× |
| from_probs | fp32 | 32 | 256k | 0.235 | 0.116 | 2.03× |

### RTX PRO 6000 (SM120)

| API | dtype | bs | vocab | before | after | speedup |
|---|---|---|---|---|---|---|
| from_logits | bf16 | 1 | 128k | 0.096 | 0.042 | 2.29× |
| from_logits | bf16 | 8 | 128k | 0.111 | 0.044 | 2.52× |
| from_logits | bf16 | 32 | 128k | 0.143 | 0.054 | 2.65× |
| from_logits | bf16 | 256 | 128k | 0.454 | 0.173 | 2.62× |
| from_logits | bf16 | 1 | 256k | 0.159 | 0.042 | 3.79× |
| from_logits | bf16 | 32 | 256k | 0.232 | 0.081 | 2.86× |
| from_logits | fp32 | 1 | 128k | 0.101 | 0.041 | 2.46× |
| from_logits | fp32 | 8 | 128k | 0.126 | 0.045 | 2.80× |
| from_logits | fp32 | 32 | 128k | 0.162 | 0.079 | 2.05× |
| from_logits | fp32 | 256 | 128k | 0.631 | 0.287 | 2.20× |
| from_logits | fp32 | 1 | 256k | 0.171 | 0.042 | 4.07× |
| from_logits | fp32 | 32 | 256k | 0.287 | 0.113 | 2.54× |
| from_probs | fp32 | 1 | 128k | 0.056 | 0.043 | 1.30× |
| from_probs | fp32 | 8 | 128k | 0.061 | 0.048 | 1.27× |
| from_probs | fp32 | 32 | 128k | 0.110 | 0.083 | 1.33× |
| from_probs | fp32 | 256 | 128k | 0.405 | 0.293 | 1.38× |
| from_probs | fp32 | 1 | 256k | 0.080 | 0.044 | 1.82× |
| from_probs | fp32 | 32 | 256k | 0.252 | 0.115 | 2.19× |

### Commands to reproduce results above
```
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_logits --batch_size 1   --vocab_size 128000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first --input_dtype bfloat16
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_logits --batch_size 8   --vocab_size 128000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first --input_dtype bfloat16
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_logits --batch_size 32  --vocab_size 128000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first --input_dtype bfloat16
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_logits --batch_size 256 --vocab_size 128000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first --input_dtype bfloat16
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_logits --batch_size 1   --vocab_size 256000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first --input_dtype bfloat16
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_logits --batch_size 32  --vocab_size 256000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first --input_dtype bfloat16
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_logits --batch_size 1   --vocab_size 128000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first --input_dtype float32
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_logits --batch_size 8   --vocab_size 128000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first --input_dtype float32
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_logits --batch_size 32  --vocab_size 128000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first --input_dtype float32
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_logits --batch_size 256 --vocab_size 128000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first --input_dtype float32
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_logits --batch_size 1   --vocab_size 256000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first --input_dtype float32
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_logits --batch_size 32  --vocab_size 256000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first --input_dtype float32
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_probs --batch_size 1   --vocab_size 128000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_probs --batch_size 8   --vocab_size 128000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_probs --batch_size 32  --vocab_size 128000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_probs --batch_size 256 --vocab_size 128000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_probs --batch_size 1   --vocab_size 256000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first
python flashinfer_benchmark.py --routine top_k_top_p_sampling_from_probs --batch_size 32  --vocab_size 256000 --top_k 50 --top_p 0.9 --filter_apply_order top_k_first
```

## 🔍 Related Issues

#3389

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [ ] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [ ] I have installed the hooks with `pre-commit install`.
- [ ] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [ ] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **Chores**
  * Added an automatic fast path that speeds up sampling when using top‑k‑first filtering with a scalar k on large vocabularies. Sampling now selects and renormalizes a smaller candidate set for faster draws, preserves determinism/tie‑breaking and validity flags when requested, falls back to the original flow when not applicable, and introduces no public API changes.

### CODE REVIEW

File: flashinfer/sampling.py [hunk 1]
```diff
@@ -1296,6 +1296,99 @@ def min_p_sampling_from_probs(
...
+
+    # top-k's default (vectorized radix) path is already CUDA-graph-capturable in
+    # deterministic mode, so we let it vectorize even during capture. Only the
+    # non-deterministic cluster path needs the graph-safe fallback (FilteredTopK with
+    # VEC_SIZE=1, ~3x slower), so request it only when actually capturing AND
+    # non-deterministic -- forcing it unconditionally regresses the common path.
+    dsa_graph_safe = not deterministic and torch.cuda.is_current_stream_capturing()
+    values, gathered_indices = _radix_top_k(
+        x, top_k, sorted=True, deterministic=deterministic, dsa_graph_safe=dsa_graph_safe
+    )
```
[the remark this answers is not in the document]
Comment (bkryu) [reply flashinfer/sampling.py]: Agree. Applied ruff formatting in next commit

File: flashinfer/sampling.py [hunk 2]
```diff
@@ -1296,6 +1296,99 @@ def min_p_sampling_from_probs(
...
+        offset=offset,
+        return_valid=return_valid,
+    )
+
+    def _map(local):
+        return (
+            gathered_indices.gather(1, local.view(-1, 1).long())
+            .squeeze(1)
+            .to(torch.int32)
+        )
```
[the remark this answers is not in the document]
Comment (bkryu) [reply flashinfer/sampling.py:1412]: Thanks for flagging this — I dug into it, but I don't think `_map` needs the change, because `top_p_sampling_from_probs` doesn't return `-1`.

In the degenerate "no valid id" case, the kernel writes a valid-range id, not a sentinel:

```cpp
// include/flashinfer/sampling.cuh, TopPSamplingFromProbKernel
if (temp_storage.last_valid_id == -1) {
  if (tx == 0) {
    output[bx] = 0;        // 0, not -1
    valid[bx] = false;
  }
  return;
}
```

In the normal case output[bx] = sampled_id ∈ [0, d). The -1 you saw (last_valid_id) is an internal scratch sentinel that's checked before the write  and never reaches output. The Python wrapper returns samples/valid verbatim (no -1, where, or masked_fill). So in _map, local is always in [0, k)  and the gather indexing is always valid — clamp(min=0) and torch.where(local >= 0, …, -1) would be dead code.
  
There's also a behavioral reason not to add it: this fast path must stay equivalent to the unfused top_k_first path, which is exactly top_p_sampling_from_probs(...) and returns 0 (flagged by valid=False) in the degenerate case. Emitting -1 instead would make the fast path diverge  from the reference path.

The real invalid signal is the valid boolean, and we already propagate it: when return_valid=True, _map returns (mapped, valid). (For the normalized  probs_k we build via softmax/renorm, this case can't realistically trigger anyway, since the running aggregate reaches 1 ≥ u.)

**Review by yzh119 — COMMENTED** [review]
One thing worth noting is that the current radix top-k implementation is not deterministic when there are tie elements (and there original implementation is deterministic)

Would be great to flag this implementation as "non-deterministic".
In the future it will be great to add deterministic radix top-k implementation.

**Comment by yzh119** [comment]
Another action item is to implement sampling kernels which accepts "index" as input so that we can fuse index remapping with sampling.

**Comment by bkryu** [comment]
> One thing worth noting is that the current radix top-k implementation is not deterministic when there are tie elements (and there original implementation is deterministic)
> Would be great to flag this implementation as "non-deterministic".
In the future it will be great to add deterministic radix top-k implementation.".

Right, thanks @yzh119 . I actually noticed that the `top_k_top_p_sampling_from_logits/from_probs` have `deterministic` input flags. To respect the flag, I've pushed e1b401c3b2f3e935004a6a9c33c8cefd966d7369 that uses the `tie_break` for the radix top-k. The fast path now passes `tie_break=SMALL` to `top_k` when `deterministic=True`. `deterministic=False` keeps the faster `tie_break=NONE`.

There is a small performance cost to `deterministic=True`, but still substantially improves vs. the existing implementation.

**Review by kahyunnam — APPROVED** [review]
